pkg/test/framework/components/echo/native/service.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package native
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"text/template"
+
+	"istio.io/istio/galley/pkg/metadata"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/galley"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+)
+
+const (
+	serviceEntryYAML = `
+apiVersion: networking.istio.io/v1alpha3
+kind: ServiceEntry
+metadata:
+  name: {{.ServiceName}}
+  labels:
+    app: {{.ServiceName}}
+    version: {{.Version}}
+spec:
+   hosts:
+   - {{.ServiceName}}.{{.Namespace}}.{{.Domain}}
+   addresses:
+   - 127.0.0.1/32
+   ports:
+   {{ range $i, $p := .Ports -}}
+   - number: {{$p.ServicePort}} 
+     name: {{$p.Name}}
+     protocol: {{$p.Protocol}}
+   {{ end -}}
+   resolution: STATIC
+   location: MESH_INTERNAL
+   endpoints:
+    - address: 127.0.0.1
+      ports:
+        {{ range $i, $p := .Ports -}}
+        {{$p.Name}}: {{$p.ServicePort}}
+        {{ end -}}
+`
+)
+
+var (
+	serviceEntryTemplate *template.Template
+
+	serviceEntryCollection = metadata.IstioNetworkingV1alpha3Serviceentries.Collection.String()
+)
+
+func init() {
+	serviceEntryTemplate = template.New("service_entry")
+	if _, err := serviceEntryTemplate.Parse(serviceEntryYAML); err != nil {
+		panic("unable to parse Service Entry template")
+	}
+}
+
+type serviceConfig struct {
+	ns      namespace.Instance
+	ports   []echo.Port
+	service string
+	domain  string
+	version string
+}
+
+func (c serviceConfig) applyTo(g galley.Instance) (*galley.SnapshotObject, error) {
```

Comment:
Are you asking why this code is adding the ServiceEntry to Galley? It seems pretty essential, since that's how Pilot discovers services.  I must be missing something?

Reply:
Sorry my comment wasn't very clear. I was curious what would happen if we didn't apply the service entry, and it seems the test passes. This suggests maybe the echo apps aren't using of the config from Pilot and are communicating just one the static config we provide, which doesn't test much? I'll have to take another look at the config. It would he great if envoy had some debug mode that tells you what routes and listeners etc matched a request to make this kind of thing clear

Reply:
Just applied some fixes. Things seem to behaving properly now.